packages/kit/src/runtime/form-utils.js

```diff
@@ -64,6 +66,202 @@ export function convert_formdata(data) {
+	const reader = request.body.getReader();
+
+	const first_chunk = await reader.read();
+	if (first_chunk.done) {
+		throw new Error('Could not deserialize binary form: empty body');
+	}
+	if (first_chunk.value.byteLength < 1 + 4 + 4) {
+		throw new Error('Could not deserialize binary form: first chunk was too small');
+	}
+	const version = first_chunk.value[0];
+	if (version !== BINARY_FORM_VERSION) {
+		throw new Error(
+			`Could not deserialize binary form: got version ${version}, expected version ${BINARY_FORM_VERSION}`
+		);
+	}
+	const start_view = new DataView(first_chunk.value.buffer);
+	const header_length = start_view.getUint32(1, true);
+	const file_count = start_view.getUint32(5, true);
+
+	// Read the header
+	const header_buffer = new Uint8Array(header_length);
+	header_buffer.set(first_chunk.value.subarray(9, header_length + 9));
+	let received_length = first_chunk.value.byteLength - 9;
+	/** @type {Array<Uint8Array>} */
+	let file_data;
+	if (received_length >= header_length) {
+		file_data = [first_chunk.value.subarray(header_length)];
+	} else {
+		while (true) {
+			const chunk = await reader.read();
+			if (chunk.done) {
+				throw new Error('Could not deserialize binary form: incomplete header');
+			}
+			const header_chunk = chunk.value.subarray(0, header_length - received_length);
+			header_buffer.set(header_chunk, received_length);
+
+			received_length += chunk.value.byteLength;
+
+			if (received_length >= header_length) {
+				file_data = [chunk.value.subarray(header_length)];
+				break;
+			}
+		}
+	}
+
+	/** @type {Array<number>} */
+	const file_sizes = new Array(file_count);
+	/** @type {Array<{start: number, end: number}> | null} */
+	let file_offsets = null;
+	/** @type {Array<Uint8Array<ArrayBuffer>>} */
+	const file_buffers = new Array(file_count);
+
+	const [data, meta] = devalue.parse(text_decoder.decode(header_buffer), {
+		File: ([name, type, size, last_modified, index]) => {
+			file_sizes[index] = size;
+			console.log(file_sizes);
+			return new Proxy(
+				new LazyFile(name, type, size, last_modified, async () => {
+					if (file_buffers[index]) return file_buffers[index];
+					if (file_offsets === null) {
+						file_offsets = new Array(file_count);
+						let start = 0;
+						for (let i = 0; i < file_count; i++) {
+							const end = start + file_sizes[i];
+							file_offsets[i] = { start, end };
+							start = end;
+						}
+					}
+					const { start, end } = file_offsets[index];
+					const buffer = new Uint8Array(end - start);
+					// let offset = 0;
+					// while (offset < buffer.byteLength) {
+					// TODO:
+					// - find the element from `file_data` that contains start + offset
+					// - if it doesn't exist, read from the request body until we get it, and cache results in `file_data`
+					// - copy subarray into `buffer`
+					// }
+					return buffer;
+				}),
+				{
+					getPrototypeOf() {
+						// Trick validators into thinking this is a normal File
+						return File.prototype;
```

Comment:
@Rich-Harris not sure the best way to do this part - we can't use actual `File` objects because that would require the data to already be buffered, and setting `__proto__ = File.prototype` causes private class member issues.